src/shogun/base/SGObject.h

```diff
@@ -370,13 +394,19 @@ class CSGObject
 		}
 	}
 
-	/** Untyped setter for an object class parameter, identified by a name.
-	 * Will attempt to convert passed object to appropriate type.
+	/** Typed setter for an object class parameter of a Shogun base class type,
+	 * identified by a name.
 	 *
 	 * @param name name of the parameter
 	 * @param value value of the parameter
 	 */
-	void put(const std::string& name, CSGObject* value);
+	template <class T,
+		      class X = typename std::enable_if<is_sg_base<T>::value>::type,
+		      class Z = void>
```

Comment:
why the `Z`?

Reply:
because SWIG confuses it with the double templated put (for non-objects) and fails. 
`
%template(put) CSGObject::put<CMachine, CMachine, void>;
`